docs/Timelines.md

```diff
@@ -0,0 +1,149 @@
+Timelines
+---------
+
+This document covers Timeline internals. It's meant to explain how
+timelines work so to help you understand the code and what each part
+of it contributes to the overall picture.
+
+# Fake timeline: your first timeline
+
+Previous to timelines, Nakadi would connect to a single Kafka cluster,
+which used to be specified in the application yaml file. So this was a
+static configuration loaded during boot time. Once timelines were
+introduced, in order to easy the migration process and remove special
+cases from the implementation, a fake timeline was introduced.
+
+Fake timelines exists only in memory. It's a Timeline which is not
+stored in the database. Fake timelines can be seen as a fallback
+strategy for when there is no actual timeline for a given event type.
+
+Also, fake timelines use the default storage:the on defined in the
+application configuration.
+
+# Timeline creation
+
+Timeline creation is coordinated through a series of locks and
+barriers using Zookeeper. Following we depict an example on how ZK
+datastructure looks like at each step.
+
+## Initial state
+
+Every time a Nakadi application is launched, it trys to create the
+following ZK structure:
+
+```
+- timelines
+  + - lock                    lock for timeline versions synchronization
+  + - version: {version}      monotonically incremented long value (version of timelines configuration)
+  + - locked_et:
+  + - nodes                   nakadi nodes
+    + - {node1}: {version}    Each nakadi node exposes version being used on this node
+    + - {node2}: {version}
+```
+
+In order not to override it, each instance needs to take the lock
+`/nakadi/timelines/lock` before executing.
+
+## Start timeline creation for et_1
+
+When a new timeline creation is initiated, the first step it to
+acquire a lock to update timelines for et_1 by creating an ephemeral
+node at `/timelines/locked_et/et_1`.
+
+```
+- timelines
+  + - lock
+  + - version: 0
+  + - locked_et:
+    + - et_1
+  + - nodes
+    + - node1: 0
+    + - node2: 0
+```
+
+## Notify all Nakadi nodes about change: the version barrier
+
+Next, it bumps the version node, which all Nakadi instances are
+listening for changes, so that they get to be aware that something is
```

Comment:
`listening for changes,` -> `listening to,`
`so that they get to be aware that something is changing` -> `so they are notified when something changes`